### DISCUSSION TITLE
How to use ortools to solve quadratic programming in Python?

### DISCUSSION THREAD

**Discussion by sato-cd** [google/or-tools#3911] [Q&A] [answered]
I want to solve Mixed-Integer Nonlinear Programming (MINLP), for example quadratic assignment problem (QAP).
 However, there is no sample code in python as far as I can see.
 Please tell me how to use ortools to solve quadratic programming in Python?

**Comment by lperron** [comment]
CP-SAT can solve qap. There is an C++ example: https://github.com/google/or-tools/blob/stable/examples/cpp/qap_sat.cc

As both C++ and python API are very similar. You can translate it to python easily.

**Comment by sato-cd** [comment]
@lperron : Thanks. I read the C++ example.

In the code, by using auxiliary variable "product", the objective is linear in itself.
Conversely, constraints are quadratic as below. (line 87-88)
> product = cp_model.NewBoolVar();
> cp_model.AddMultiplicationEquality(product, place_vars[f1][l1], place_vars[f2][l2]);

I understand that we can handle QAP by using auxiliary variable and quadratic constraints.
Is my understanding correct? I want to make sure.
Also, is there any other way to handle QAP?  For example, the objective is quadratic in itself.

**Comment by lperron** [comment, marked as answer]
just add the target variable of the AddMultiplicationEquality() to the linear objective.

**Comment by sato-cd** [comment]
@lperron :  Thank you so much.

Please let me ask you a related question. I would like to use SCIP as solver if possible.
In using SCIP, is it the same coding as above? Or is it different?

**Comment by lperron** [comment]
Different, and we currently have no api to set the quadratic part.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00
